lib/LDAPUserManager.php

```diff
@@ -350,30 +335,75 @@ public function deleteUser($uid): bool {
 		return $res;
 	}
 
+	/**
+	 * checks whether the user is allowed to change their password in Nextcloud
+	 *
+	 * @return boolean either the user can or cannot
+	 */
+	public function canSetPassword(): bool {
+		return $this->configuration->hasPasswordPermission();
+	}
+
 	/**
 	 * Set password
 	 *
 	 * @param string $uid The username
 	 * @param string $password The new password
+	 * @param ?\LDAP\Connection $connection LDAP connection or null to create one
 	 * @return bool
 	 *
 	 * Change the password of a user
 	 */
-	public function setPassword($uid, $password) {
-		if (!function_exists('ldap_exop_passwd')) {
-			// since PHP 7.2 – respondToActions checked this already, this
-			// method should not be called. Double check due to public scope.
-			// This method can be removed when Nextcloud 16 compat is dropped.
-			return false;
+	public function setPassword($uid, $password, $connection = null) {
+		if (is_null($connection)) {
+			$connection = $this->ldapProvider->getLDAPConnection($uid);
 		}
+		
 		try {
-			$cr = $this->ldapProvider->getLDAPConnection($uid);
 			$userDN = $this->getUserDN($uid);
-			return ldap_exop_passwd($cr, $userDN, '', $password) !== false;
+			$ret = false;
+
+			// try ldap_exop_passwd first
+			if ($this->ldapConnect->hasPasswdExopSupport($connection)) {
+				$ret = ldap_exop_passwd($connection, $userDN, '', $password);
+				if ($ret === false) {
+					$message = 'Failed to set password for user {dn} using ldap_exop_passwd';
+					$this->logger->error($message, [
+						'ldap_error' => ldap_error($connection),
+						'app' => Application::APP_ID,
+						'dn' => $userDN,
+					]);
+				} else {
+					// `ldap_exop_passwd` is either FALSE or the password, in the later case return TRUE
+					$ret = true;
+				}
+			} else {
+				// Fallback to `userPassword` in case the server does not support exop_passwd
+				$entry = [];
+				if ($this->configuration->useUnicodePassword()) {
+					$entry['unicodePwd'] = iconv('UTF-8', 'UTF-16LE', '"' . $password . '"');
+				} else {
+					$entry['userPassword'] = $password;
```

Comment:
May want to hash this? But then it opens a can of worms of which hash to use and so on…

Reply:
There are implementations hashing it automatically if not hashed already, so not sure if we want to go into that rabbit hole of choosing a hash algorithm.